JBPM-7750: Create Work Item for executing sql statements

jesuino opened the pull request:
JBPM-7750: Create Work Item for fetching data from database

tsurdilo commented:
@jesuino thanks! i will look over it asap.

tsurdilo commented:
@jesuino I updated your pr directly and made some changes to the handler and tests and updated some poms so that your new handlers is part of the workitem repository. I also renamed it from "fetch data" to "execute sql". Can you please have a look at the updated code and let me know what you think. 
Thank you for the great contribution!

@mswiderski can you please take a look? Thanks

jesuino commented:
> @jesuino updated your pr directly and made some changes to the handler and tests and updated some poms so that your new handlers is part of the workitem repository. I also renamed it from "fetch data" to "execute sql". Can you please have a look at the updated code and let me know what you think.
> Thank you for the great contribution!


Hello @tsurdilo thanks for your analysis, I think it looks great, thanks!

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,130 @@
+                keywords = "database,fetch,sql,execute",
+                action = @WidAction(title = "Execute SQL statements")
+        ))
+public class ExecuteSqlWorkItemHandler extends AbstractLogOrThrowWorkItemHandler {
```
I would recommend to implement Cacheable interface for better performance

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+            List<String> lines = new ArrayList<>();
+            try {
+                DataSource ds = InitialContext.doLookup(dataSource);
```
move this up from the execute method into constructor so the data source look up will be done only once. If you think that single handler should use multiple data sources then I'd make it an option instead of default.

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,130 @@
+                statement = connection.prepareStatement(sqlStatement);
+                statement.setMaxRows(maxResults != null && !maxResults.trim().isEmpty() ? Integer.parseInt(maxResults) : DEFAULT_MAX_RESULTS);
+                resultSet = statement.executeQuery();
+                while (resultSet.next()) {
```
maybe create a method (protected so it can be overridden) for processing result set

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,130 @@
+                        connection.close();
+                    }
+                } catch (SQLException e) {
+                    handleException(e);
```
not sure we should handle exception on closing resources

tsurdilo commented:
@mswiderski updated . thanks.

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,163 @@
+
+            List<String> lines = new ArrayList<>();
+            try {
+                ds = InitialContext.doLookup(dataSourceName);
```
move this line (InitialContext) to constructor

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,163 @@
+    private static final String DEFAULT_COLUMN_SEPARATOR = ",";
+    private ResultSet resultSet;
+    private Connection connection;
+    private PreparedStatement statement;
```
connection, prepared statement and resultset should be created and closed within the execute method

mswiderski commented on execute-sql-workitem/src/main/java/org/jbpm/process/workitem/executesql/ExecuteSqlWorkItemHandler.java:
```diff
@@ -0,0 +1,163 @@
+                statement.close();
+            }
+            if (connection != null) {
+                connection.close();
```
no need for this as these instances should be in execute method only

tsurdilo commented:
@mswiderski updated

jesuino commented:
@mswiderski Yes, thanks again for your patience and advice, master Maciej!
